File: WQChartsDemo/Objc/View/SliderCell.m
// 代码地址: https://github.com/CoderWQYao/WQCharts-iOS
//
// SliderCell.m
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

#import "SliderCell.h"

@interface SliderCell()

@property (nonatomic, strong) UISlider* slider;
@property (nonatomic, strong) UILabel* textLabel;
@property (nonatomic, strong) UILabel* placeholderLabel;

@end

@implementation SliderCell

#pragma mark - Public

- (instancetype)initWithFrame:(CGRect)frame {
    if(self = [super initWithFrame:frame]) {
        self.backgroundColor = UIColor.clearColor;
        [self addSubview:self.slider];
        [self addSubview:self.textLabel];
    }
    return self;
}


- (void)layoutSubviews {
    [super layoutSubviews];
    
    CGFloat width = self.bounds.size.width;
    CGFloat height = self.bounds.size.height;

    UILabel* textLabel = self.textLabel;
    UILabel* placeholderLabel = self.placeholderLabel;
    [placeholderLabel sizeToFit];
    CGSize textSize = placeholderLabel.bounds.size;
    textSize.width += 6;
    textLabel.frame = CGRectMake(width - textSize.width - 5, (height - textSize.height) / 2, textSize.width, textSize.height);
    
    UISlider* slider = self.slider;
    [slider sizeToFit];
    slider.frame = CGRectMake(5, (height - slider.bounds.size.height) / 2, CGRectGetMinX(textLabel.frame) - 5, slider.bounds.size.height);
}


- (CGSize)sizeThatFits:(CGSize)size {
    return CGSizeMake(size.width, MIN(size.height, 40));
}

- (void)callSliderValueChange {
    if(self.onValueChange) {
        self.onValueChange(self, self.value);
      }
}

#pragma mark - Getter

- (UISlider *)slider {
    if(!_slider) {
        _slider = [[UISlider alloc] init];
        _slider.tintColor = Color_Tint;
        [_slider addTarget:self action:@selector(onSliderValueChange) forControlEvents:UIControlEventValueChanged];
    }
    return _slider;
}

- (UILabel *)textLabel {
    if(!_textLabel) {
        _textLabel = [[UILabel alloc] init];
        _textLabel.font = [UIFont systemFontOfSize:9];
        _textLabel.textAlignment = NSTextAlignmentRight;
        _textLabel.textColor = Color_White;
    }
    return _textLabel;
}

- (UILabel *)placeholderLabel {
    if(!_placeholderLabel) {
        _placeholderLabel = [[UILabel alloc] init];
        _placeholderLabel.font = self.textLabel.font;
        _placeholderLabel.textAlignment = NSTextAlignmentRight;
    }
    return _placeholderLabel;
}

- (float)value {
    return _slider.value;
}

- (float)minimumValue {
    return _slider.minimumValue;
}

- (float)maximumValue {
    return _slider.maximumValue;
}

#pragma mark Setter

- (void)setValue:(float)value {
    _slider.value = value;
    [self updateTextLabel];
}

- (void)setMinimumValue:(float)minimumValue {
    _slider.minimumValue = minimumValue;
    [self updateTextLabel];
}

-(void)setMaximumValue:(float)maximumValue {
    _slider.maximumValue = maximumValue;
    [self updateTextLabel];
}


- (SliderCell*(^)(float,float,float))setSliderValue {
    return ^SliderCell *(float minimumValue,float maximumValue,float value) {
        UISlider* slider = self.slider;
        slider.minimumValue = minimumValue;
        slider.maximumValue = maximumValue;
        slider.value = value;
        [self updateTextLabel];
        return self;
    };
}

- (void)setDecimalCount:(NSInteger)decimalCount {
    _decimalCount = decimalCount;
    [self updateTextLabel];
}

- (SliderCell*(^)(NSInteger))setDecimalCount {
    return ^SliderCell *(NSInteger decimalCount) {
        self.decimalCount = decimalCount;
        return self;
    };
}

- (SliderCell*(^)(SliderCellOnValueChange))setOnValueChange {
    return ^SliderCell *(SliderCellOnValueChange onValueChange) {
        self.onValueChange = onValueChange;
        return self;
    };
}

- (SliderCell*(^)(id))setObject {
    return ^SliderCell *(id objcect) {
        self.object = objcect;
        return self;
    };
}

#pragma mark Private

- (NSString*)createValueStringWithValue:(float)value {
    NSString* format = [NSString stringWithFormat:@"%%.%ldf",self.decimalCount];
    return [NSString stringWithFormat:format,self.decimalCount == 0 ? round(value) : value];
 }

- (void)updateTextLabel {
    UISlider* slider = self.slider;
    self.textLabel.text = [self createValueStringWithValue:slider.value];
    self.placeholderLabel.text = [NSString stringWithFormat:@"-%@",[self createValueStringWithValue:MAX(ABS(slider.minimumValue),ABS(slider.maximumValue))]];
    [self setNeedsLayout];
}

- (void)onSliderValueChange {
    UILabel* textLabel = self.textLabel;
    NSString* oldString = textLabel.text;
    [self updateTextLabel];
    NSString* newString = textLabel.text;
    if(![oldString isEqualToString:newString]) {
        [self callSliderValueChange];
    }
}

@end

File: WQChartsDemo/Swift/View/SliderCell.swift
// 代码地址: 
// SliderCell.swift
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

import UIKit

class SliderCell: UIView {
    
    // MARK: - Property
    
    typealias OnValueChange = (_ sliderCell: SliderCell, _ value: CGFloat) -> Void
    
    open lazy var textLabel: UILabel = {
        let textLabel = UILabel()
        textLabel.font = UIFont.systemFont(ofSize: 9)
        textLabel.textAlignment = .right
        textLabel.textColor = Color_White
        return textLabel
    }()
    
    open lazy var slider: UISlider = {
        let slider = UISlider()
        slider.addTarget(self, action: #selector(onSliderValueChange), for: .valueChanged)
        return slider
    }()
    
    open lazy var placeholderLabel: UILabel = {
        let placeholderLabel = UILabel()
        placeholderLabel.font = textLabel.font
        placeholderLabel.textAlignment = .right
        return placeholderLabel
    }()
    
    open var value: CGFloat {
        get {
            return CGFloat(Float(textLabel.text ?? "") ?? 0)
        }
        set {
            slider.value = Float(newValue)
            updateTextLabel()
        }
    }
    
    open var minimumValue: CGFloat {
        get {
            return CGFloat(slider.minimumValue)
        } set {
            slider.minimumValue = Float(newValue)
            updateTextLabel()
        }
    }
    
    open var maximumValue: CGFloat {
        get {
            return CGFloat(slider.maximumValue)
        } set {
            slider.maximumValue = Float(newValue)
            updateTextLabel()
        }
    }
    
    open func setSliderValue(_ minimumValue: CGFloat, _ maximumValue: CGFloat, _ value: CGFloat) -> SliderCell {
        let slider = self.slider
        slider.minimumValue = Float(minimumValue)
        slider.maximumValue = Float(maximumValue)
        slider.value = Float(value)
        updateTextLabel()
        return self
    }
    
    open var decimalCount = Int(0) {
        didSet {
            updateTextLabel()
        }
    }
    
    open func setDecimalCount(_ decimalCount: Int) -> SliderCell {
        self.decimalCount = decimalCount
        return self
    }
    
    open var onValueChange: OnValueChange?
    
    open func setOnValueChange(_ onValueChange: OnValueChange?) -> SliderCell {
        self.onValueChange = onValueChange
        return self
    }
    
    public var object: Any?
    
    public func setObject(_ object: Any?) -> SliderCell {
        self.object = object
        return self
    }

    // MARK: - Public
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .clear
        addSubview(slider)
        addSubview(textLabel)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        let width = bounds.width
        let height = bounds.height
        
        let textLabel = self.textLabel
        let placeholderLabel = self.placeholderLabel
        placeholderLabel.sizeToFit()
        var textSize = placeholderLabel.bounds.size
        textSize.width += 6
        textLabel.frame = CGRect(x: width - textSize.width - 5, y: (height - textSize.height) / 2, width: textSize.width, height: textSize.height)
        
        let slider = self.slider
        slider.sizeToFit()
        slider.frame = CGRect(x: 5, y: (height - slider.bounds.height) / 2, width: textLabel.frame.minX - 5, height: slider.bounds.height)
    }
    
    override func sizeThatFits(_ size: CGSize) -> CGSize {
        return CGSize(width: size.width, height: min(size.height,40))
    }
    
    open func callSliderValueChange() {
        if let onValueChange = onValueChange {
            onValueChange(self,value)
        }
    }
    
    // MARK: - Private
    
    func createValueString(_ value: Float) -> String {
        let format = String(format: "%%.%ldf", decimalCount)
        return String(format: format, decimalCount == 0 ? round(value) : value)
    }
    
    func updateTextLabel() {
        let slider = self.slider
        textLabel.text = createValueString(slider.value)
        placeholderLabel.text = String(format: "-%@", createValueString(max(abs(slider.minimumValue),abs(slider.maximumValue))))
        setNeedsLayout()
    }
    
    @objc func onSliderValueChange()  {
        let textLabel = self.textLabel
        let oldString = textLabel.text
        updateTextLabel()
        let newString = textLabel.text
        if oldString != newString {
            callSliderValueChange()
        }
    }
    
    
}
